libass/ass_render.c

```diff
@@ -2201,19 +2218,13 @@ static void render_and_combine_glyphs(ASS_Renderer *render_priv,
             if (flags == FILTER_NONZERO_SHADOW && (info->c[0] & 0xFF) == 0xFF)
                 flags = 0;
 
-            ASS_Vector pos = {0}, pos_o = {0};
-            info->pos.x = double_to_d6(device_x + d6_to_double(info->pos.x) * render_priv->font_scale_x);
-            info->pos.y = double_to_d6(device_y) + info->pos.y;
-            get_bitmap_glyph(render_priv, info, &pos, &pos_o, flags);
-
             if (linebreak || is_new_bm_run(info, last_info)) {
                 linebreak = 0;
                 last_info = NULL;
                 if (nb_bitmaps >= text_info->max_bitmaps) {
                     size_t new_size = 2 * text_info->max_bitmaps;
                     if (!ASS_REALLOC_ARRAY(text_info->combined_bitmaps, new_size)) {
-                        ass_cache_dec_ref(info->bm);
-                        ass_cache_dec_ref(info->bm_o);
+                        ass_cache_dec_ref(info->outline);
```

Comment:
No, not here. I'm moving `get_bitmap_glyph()` down so in this place I should free outline instead of bitmap.

`get_bitmap_glyph()` always consumes outlines: either directly through `ass_cache_dec_ref()` or indirectly through `ass_cache_get()` which transfers ownership of outline to bitmap cache key.

Reply:
> get_bitmap_glyph() always consumes outlines: either directly through ass_cache_dec_ref() or indirectly through ass_cache_get() which transfers ownership of outline to bitmap cache key.

OK, that’s not intuitive at all. I’ll look through your `ass_cache_dec_ref` calls again.